XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Core
+{
+	public class StreamMediaSource : MediaSource, IStreamImageSource
+	{
+		readonly object synchandle = new object();
+		CancellationTokenSource cancellationTokenSource;
+
+		TaskCompletionSource<bool> completionSource;
+
+		public static readonly BindableProperty StreamProperty = BindableProperty.Create(nameof(Stream), typeof(Func<CancellationToken, Task<Stream>>), typeof(StreamMediaSource),
+			default(Func<CancellationToken, Task<Stream>>));
+
+		protected CancellationTokenSource CancellationTokenSource
+		{
+			get => cancellationTokenSource;
+			private set
+			{
+				if (cancellationTokenSource == value)
+					return;
+				if (cancellationTokenSource != null)
+				{
+					cancellationTokenSource.Cancel();
+					cancellationTokenSource.Dispose();
+				}
+				cancellationTokenSource = value;
+			}
+		}
+
+		bool IsLoading => cancellationTokenSource != null;
+
+		public virtual Func<CancellationToken, Task<Stream>> Stream
+		{
+			get => (Func<CancellationToken, Task<Stream>>)GetValue(StreamProperty);
+			set => SetValue(StreamProperty, value);
+		}
+
+		protected override void OnPropertyChanged(string propertyName)
+		{
+			if (propertyName == StreamProperty.PropertyName)
+				OnSourceChanged();
+			base.OnPropertyChanged(propertyName);
+		} 
+
+		async Task<Stream> IStreamImageSource.GetStreamAsync(CancellationToken userToken)
+		{
+			if (Stream == null)
+				return null;
+
+			OnLoadingStarted();
+			userToken.Register(CancellationTokenSource.Cancel);
+			try
+			{
+				var stream = await Stream(CancellationTokenSource.Token);
+				OnLoadingCompleted(false);
+				return stream;
+			}
+			catch (OperationCanceledException)
+			{
+				OnLoadingCompleted(true);
+				throw;
+			}
+		}
+
+		protected void OnLoadingCompleted(bool cancelled)
+		{
+			if (!IsLoading || completionSource == null)
+				return;
+
+			var tcs = Interlocked.Exchange(ref completionSource, null);
+			if (tcs != null)
+				tcs.SetResult(cancelled);
+
+			lock (synchandle)
```

Comment:
I guess this lock should be moved inside of CancellationTokenSource setter, no?

Reply:
@AndreiMisiukevich, do you have a code snippet for that?

Reply:
Since it all works, I'm going to keep it like this for a first version. But we can definitely experiment with this.

Reply:
Also, still not sure this object is even used.